stock-managment/src/main/java/org/dhis2/android/rtsm/ui/managestock/ManageStockViewModel.kt

```diff
@@ -245,6 +266,131 @@ class ManageStockViewModel @Inject constructor(
         )
     )
 
+    fun onCellValueChanged(tableCell: TableCell) {
+        val updatedData = screenState.value?.tables?.map { tableModel ->
+            if (tableModel.hasCellWithId(tableCell.id)) {
+                tableModel.copy(
+                    overwrittenValues = mapOf(
+                        Pair(tableCell.column!!, tableCell)
+                    )
+                )
+            } else {
+                tableModel
+            }
+        } ?: emptyList()
+
+        _screenState.postValue(
+            TableScreenState(
+                tables = updatedData,
+                selectNext = false
+            )
+        )
+    }
+
+    fun onCellClick(cell: TableCell): TextInputModel {
+        val stockItem = _stockItems.value?.find { it.id == cell.id }
+        val itemName = stockItem?.name ?: ""
+        return TextInputModel(
+            id = cell.id ?: "",
+            mainLabel = itemName,
+            secondaryLabels = mutableListOf(resources.getString(R.string.quantity)),
+            currentValue = cell.value,
+            keyboardInputType = KeyboardInputType.NumericInput(
+                allowDecimal = false,
+                allowSigned = false
+            ),
+            error = errors.value[cell.id!!]
+        )
+    }
+
+    fun onSaveValueChange(
+        cell: TableCell,
+        selectNext: Boolean
+    ) {
+        // When user taps on done or next. We should apply program rules here
+        val stockItem = _stockItems.value?.find { it.id == cell.id }
+        stockItem?.let {
+            cell.value?.let { value ->
+                setQuantity(
+                    it, 0, value,
+                    object : ItemWatcher.OnQuantityValidated {
+                        override fun validationCompleted(ruleEffects: List<RuleEffect>) {
+                            // Update fields
+                            ruleEffects.forEach { ruleEffect ->
+                                if (ruleEffect.ruleAction() is RuleActionAssign &&
+                                    (
+                                        (ruleEffect.ruleAction() as RuleActionAssign).field()
+                                            == config.value?.stockOnHand
+                                        )
+                                ) {
+                                    val data = ruleEffect.data()
+                                    val isValid: Boolean = isValidStockOnHand(data)
+                                    val stockOnHand = if (isValid) data else it.stockOnHand
+                                    addItem(it, cell.value, stockOnHand, !isValid)
+                                    if (!isValid) {
+                                        errors.value[cell.id!!] = resources
+                                            .getString(R.string.stock_on_hand_exceeded_message)
+                                    } else {
+                                        errors.value.remove(cell.id)
+                                    }
```

Comment:
I think that we can remove all this code block: 
```
if (!isValid) {
                                        errors.value[cell.id!!] = resources
                                            .getString(R.string.stock_on_hand_exceeded_message)
                                    } else {
                                        errors.value.remove(cell.id)
```
Beacuse we are already storing the error related to the item in line 329:
`addItem(it, cell.value, stockOnHand, !isValid)`
We already have if it is valid or not.

We only need to check it below...